opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,604 @@
+    auto robot_pose = getRobotPoseInFrame(target_pose.header.frame_id);
+    const double dist = std::hypot(
+      robot_pose.pose.position.x - target_pose.pose.position.x,
+      robot_pose.pose.position.y - target_pose.pose.position.y);
+    bool reversing = allow_backward_ &&
+      dist < desired_distance_ && target_pose.pose.position.x < robot_pose.pose.position.x;
+
+    // Compute and publish controls
+    auto command = std::make_unique<geometry_msgs::msg::TwistStamped>();
+    command->header.stamp = now();
+    if (!controller_->computeVelocityCommand(target_pose.pose, command->twist, true, reversing)) {
+      throw opennav_following::FailedToControl("Failed to get control");
+    }
+    vel_publisher_->publish(std::move(command));
+
+    loop_rate.sleep();
+  }
+  return false;
+}
+
+bool FollowingServer::rotateToObject(geometry_msgs::msg::PoseStamped & object_pose)
+{
+  const double dt = 1.0 / controller_frequency_;
+  auto target_pose = object_pose;
+  target_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+    tf2::getYaw(target_pose.pose.orientation) + M_PI);
+
+  rclcpp::Rate loop_rate(controller_frequency_);
+  auto start = this->now();
+  auto timeout = rclcpp::Duration::from_seconds(rotate_to_object_timeout_);
+
+  while (rclcpp::ok()) {
+    publishFollowingFeedback(FollowObject::Feedback::RETRY);
+
+    // Stop if cancelled/preempted
+    if (checkAndWarnIfCancelled<FollowObject>(following_action_server_, "follow_object") ||
+      checkAndWarnIfPreempted<FollowObject>(following_action_server_, "follow_object"))
+    {
+      return false;
+    }
+
+    auto robot_pose = getRobotPoseInFrame(object_pose.header.frame_id);
+    auto angular_distance_to_heading = angles::shortest_angular_distance(
+      tf2::getYaw(robot_pose.pose.orientation), tf2::getYaw(target_pose.pose.orientation));
+    if (fabs(angular_distance_to_heading) < angular_tolerance_) {
+      throw opennav_following::FailedToControl("Failed to rotate to object");
+    }
+
+    // Determine if we find the object
+    if (getRefinedPose(object_pose)) {
+      return true;
+    }
+
+    auto current_vel = std::make_unique<geometry_msgs::msg::TwistStamped>();
+    current_vel->twist.angular.z = odom_sub_->getTwist().theta;
+
+    auto command = std::make_unique<geometry_msgs::msg::TwistStamped>();
+    command->header = robot_pose.header;
+    command->twist = controller_->computeRotateToHeadingCommand(
+      angular_distance_to_heading, current_vel->twist, dt);
+
+    vel_publisher_->publish(std::move(command));
+
+    if (this->now() - start > timeout) {
+      throw opennav_following::FailedToControl("Timed out rotating to object");
+    }
+
+    loop_rate.sleep();
+  }
+
+  return false;
+}
+
+geometry_msgs::msg::PoseStamped FollowingServer::getRobotPoseInFrame(const std::string & frame)
+{
+  geometry_msgs::msg::PoseStamped robot_pose;
+  robot_pose.header.frame_id = base_frame_;
+  robot_pose.header.stamp = rclcpp::Time(0);
```

Comment:
Use an actual timestamp please - we should keep this as synchronzied as possible. Whatever the start timestamp is for an iteration we should use for getting all robot position, obstacle frame transformation, and control. The `stamp` used for all should be the same so we know that our robot's pose, target's pose, and controller are all working with the same snapshot in time